### PR TITLE
HDDS-7191. Separate prop for s3 admin

### PR DESCRIPTION
**Pull request by Pochatkin**
## What changes were proposed in this pull request?

1. Created new props `ozone.s3.administrators` and `ozone.s3.administrators.groups`
2. In case when at least one of these props is defined all s3 shell operation can be executed only by one of defined user as admin. Each user still should have permission to generate keys for itself.
3. In case when these properties are empty admins should be taken from `ozone.administrators` or `ozone.administrators.groups`

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7191

## How was this patch tested?

Manual testing and new unit tests

### CODE REVIEW

**Comment by Pochatkin** [comment]
@adoroszlai Could you please give approve for running workflows?

**Comment by kerneltime** [comment]
We would need a robot test as well to make sure s3 admins and ozone admins can continue to retrieve the secret keys. 
The change makes sense overall.

**Comment by Pochatkin** [comment]
@kerneltime Done. Could you please review this changes or tag anyone who can?

**Comment by myskov** [comment]
@adoroszlai could you please take a look at the PR?

**Review by adoroszlai — COMMENTED** [review]
Thanks @Pochatkin for the patch, LGTM.

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/s3/security/TestS3GetSecretRequest.java [hunk 1]
```diff
@@ -329,11 +334,48 @@ public void testGetSecretOfAnotherUserAsNonAdmin() throws IOException {
         + "the permission to get bob's secret in this test case!");
   }
 
+  @Test
+  public void testGetSecretOfAnotherUserAsOzoneAdmin() throws IOException {
+    // This effectively makes alice an admin.
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/s3/security/TestS3GetSecretRequest.java:339]:
```suggestion
    // This effectively makes alice an S3 admin.
```

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/s3/security/TestS3GetSecretRequest.java [hunk 2]
```diff
@@ -196,7 +201,7 @@ private OMRequest s3GetSecretRequest(String userPrincipalStr) {
   public void testGetSecretOfAnotherUserAsAdmin() throws IOException {
 
     // This effectively makes alice an admin.
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/s3/security/TestS3GetSecretRequest.java:203]:
```suggestion
    // This effectively makes alice an S3 admin.
```

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/s3/security/TestS3GetSecretRequest.java [hunk 3]
```diff
@@ -329,11 +334,48 @@ public void testGetSecretOfAnotherUserAsNonAdmin() throws IOException {
...
+    // The secret inside should be the same.
+    final S3Secret s3Secret1 = getS3SecretResponse.getS3Secret();
+    Assert.assertEquals(USER_CAROL, s3Secret1.getKerberosID());
+    Assert.assertEquals(awsSecret1, s3Secret1.getAwsSecret());
+  }
+
   @Test
   public void testGetSecretWithTenant() throws IOException {
 
     // This effectively makes alice an admin.
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/s3/security/TestS3GetSecretRequest.java:377]:
```suggestion
    // This effectively makes alice an S3 admin.
```

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/s3/security/TestS3GetSecretRequest.java [hunk 4]
```diff
@@ -329,11 +334,48 @@ public void testGetSecretOfAnotherUserAsNonAdmin() throws IOException {
...
+    Assert.assertNotNull(awsSecret1);
+
+    final GetS3SecretResponse getS3SecretResponse =
+        s3GetSecretResponse.getOMResponse().getGetS3SecretResponse();
+    // The secret inside should be the same.
+    final S3Secret s3Secret1 = getS3SecretResponse.getS3Secret();
+    Assert.assertEquals(USER_CAROL, s3Secret1.getKerberosID());
+    Assert.assertEquals(awsSecret1, s3Secret1.getAwsSecret());
+  }
+
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/s3/security/TestS3GetSecretRequest.java:373]: Can we add a negative test case where e.g. `alice` is an Ozone admin, but **not** an S3 admin, thus can't do S3 secret operations (rejected) for others?

**Review by smengcl — COMMENTED** [review]
Thanks for the patch and UT addition. Overall looks good.

**Comment by Pochatkin** [comment]
@smengcl Done

**Comment by adoroszlai** [comment]
@Pochatkin Can you please enable the build-branch workflow in your fork?  Also, checkstyle check is quick, can be run locally simply by `hadoop-ozone/dev-support/checks/checkstyle.sh`.

**Comment by Pochatkin** [comment]
@adoroszlai Sorry, missed checkstyle warnings in output because build was SUCCESS (this is unpredictable that checkstyle fails doesn't fail whole build).  Workflow enabled. Thank you

**Comment by adoroszlai** [comment]
> this is unpredictable that checkstyle fails doesn't fail whole build

I think most developers run checkstyle via the script `hadoop-ozone/dev-support/checks/checkstyle.sh`, which fails on violation, and prints violations in a more useful way.

We can change the default config, though, to cover the case where `mvn checkstyle:check` is run manually:

https://github.com/apache/ozone/blob/3d3e9eaf00385f2d2b5050792e96092e07cc09d7/pom.xml#L2039

That should be fine, as `hadoop-ozone/dev-support/checks/checkstyle.sh` explicitly sets it:

https://github.com/apache/ozone/blob/3d3e9eaf00385f2d2b5050792e96092e07cc09d7/hadoop-ozone/dev-support/checks/checkstyle.sh#L24

**Review by smengcl — APPROVED** [review]

**Comment by adoroszlai** [comment]
Thanks @Pochatkin for iterating on the patch, @kerneltime, @smengcl for the review.

**Pull request merged**